search: Get search-blitz ready for deployment

camdencheek opened the pull request:
This is kinda a combo PR to get search-blitz ready to deploy against sourcegraph.com.

1) Adds the sentinel queries to the config file
2) Modifies the Dockerfile so it can build
3) Adds some randomization to start times so we're not spiking every minute

Fixes #19893

camdencheek commented on internal/cmd/search-blitz/Dockerfile:
```diff
@@ -1,19 +1,17 @@
-RUN go mod download
 COPY . .
-RUN CGO_ENABLED=0 GOOS=linux go build -o searchblitz .
+RUN CGO_ENABLED=0 GOOS=linux go build -o searchblitz ./internal/cmd/search-blitz
```
I really don't like the changes I had to make to the Dockerfile.

Summary: 
- We need access to the internal streaming client
- For access to an internal package, search-blitz needs to be in the same module
- To be in the same module, it requires search-blitz share a go.mod file with sourcegraph/sourcegraph
- To share a go.mod file (and all the internal code), the Dockerfile needs to be built with the entire context of the sourcegraph/sourcegraph repo

I'm starting to think we should just make the streaming client public, but am interested to hear if there are other ideas to solve this less in a less dirty way than `docker build -f ./internal/cmd/search-blitz/Dockerfile .`

camdencheek replied:
The real cost is pretty slim. 

Build times increase significantly (copying context and downloading modules), but considering that we don't build the docker image regularly, I don't consider this a huge deal. Build size is unaffected since the build is staged. 

It also somewhat increases cognitive complexity of building it, but looking again, I should probably just update the scripts so people don't have to think about the docker context.

camdencheek replied:
It took 1m30s to copy, download modules, and build. Not awful, not great. I think it's workable for now

stefanhengl replied:
>  Not awful, not great.

That's the sweet spot :-)

camdencheek commented on internal/cmd/search-blitz/config.yaml:
```diff
@@ -1,56 +1,100 @@
+                  strings.ToUpper(...)
+            name: structural_multi_repo_small
+
+# Queries that were previously used, and may still be useful, but are larger than
```
Should we keep these queries in the file? Or just let them live in git history for now?

camdencheek replied:
They probably make most sense in the a dot-com configmap rather than in a yaml file built into the image, so I expect this is a temporary measure anyway

stefanhengl replied:
I am a bit concerned that we are loosing variability in the queries. The new queries are either global or go against sourcegraph. For example, we don't cover different number of shards or very popular repositories. The advantage of taking queries from actual users instead of hand-crafted queries is that we get a realistic measure of what users experience. Not a blocker for this PR, but it is something we have to be aware of.

stefanhengl replied:
What about global literal searches?

camdencheek replied:
> There is also something nice about having them checked in: it is harder to change it and the Graphs become more reliable.

Ooh, I hadn't thought about that. Yeah, I agree. Ideally these are largely static.

camdencheek replied:
> Not a blocker for this PR

Honestly, I'd like this to be a blocker. Your concern about the breadth of searches these cover is totally valid. I'd like to get a really solid cohort of sentinel queries together that we can feel confident about them representing what users see. Would you mind pairing with me on this for a few at some point?

camdencheek replied:
> What about global literal searches?

Good catch. I kept forgetting what default search type is. I should probably just make `patternType` explicit for every query :)

stefanhengl replied:
>Would you mind pairing with me on this for a few at some point?

Sure! You can always send me an invite or ping me when you catch me online. I saw that you improved the queries a couple of hours after this comment. They look really good now!

I wonder whether it makes sense to do something similar to what we do for integration tests IE fork test repos to a dedicated org. WDYT?

camdencheek replied:
Makes sense to me. Should help keep our result set more static

stefanhengl commented on internal/cmd/search-blitz/config.yaml:
```diff
@@ -1,56 +1,100 @@
+            name:  regex_small
+          - query: can_handle_hl_color patterntype:regexp # ~ 50 results
+            name:  regex_medium
+          - query: try_init # 500+ results
```
patternType?

camdencheek commented:
I significantly revamped the queries to cover more cases:
- Added multiple literal searches
- Changed most of the searches targeting sourcegraph repos to target the repos in the original set of queries
- Changed some of the regex queries to actually use regex (nothing super expensive though)
- Added an explicit `patterntype` to every query

camdencheek commented:
Additionally, I updated the build scripts to make building the image less weird. Takes a while at 1m30s, but I don't expect us to be building this all the time.

stefanhengl reviewed (commented):
LGTM!
